corehq/apps/app_manager/xform.py

```diff
@@ -1404,36 +1407,93 @@ def make_delegation_stub_case_block():
+
+            self.add_instance(
+                form_xpath.fixture_id,
+                u'jr://fixture/{}'.format(form_xpath.fixture_id)
+            )
+
+            self.add_bind(
+                nodeset=u'/data/{}'.format(name),
+                calculate=form_xpath.xpath_phase_set
+            )
+            self.data_node.append(_make_elem(name))
+
+        self.add_bind(
+            nodeset=u'/data/{}'.format(SCHEDULE_GLOBAL_NEXT_VISIT_DATE),
+            calculate=u'date(min({}))'.format(','.join(forms_due))
+        )
+        self.data_node.append(_make_elem(SCHEDULE_GLOBAL_NEXT_VISIT_DATE))
+
     def create_casexml_2_advanced(self, form):
         from corehq.apps.app_manager.util import split_path
 
-        def configure_visit_schedule_updates(update_block):
+        if not form.actions.get_all_actions():
+            return
+
+        case_tag = lambda a: "case_{0}".format(a.case_tag)
+
+        def configure_visit_schedule_updates(update_block, action):
+            case = CaseIDXPath(session_var(action.case_session_var)).case()
+            schedule_form_xpath = QualifiedScheduleFormXPath(form, form.get_phase(), form.get_module(), case)
+
+            self.add_instance(
+                schedule_form_xpath.fixture_id,
+                u'jr://fixture/{}'.format(schedule_form_xpath.fixture_id)
+            )
+
+            self.add_bind(
+                nodeset=u'{}/case/update/{}'.format(case_tag(action), SCHEDULE_PHASE),
+                type="xs:integer",
+                calculate=schedule_form_xpath.current_schedule_phase_calculation(
+                    self.action_relevance(form.schedule.termination_condition),
+                    self.action_relevance(form.schedule.transition_condition),
+                )
+            )
             update_block.append(make_case_elem(SCHEDULE_PHASE))
-            last_visit_num = SCHEDULE_LAST_VISIT.format(form.schedule_form_id)
-            last_visit_date = SCHEDULE_LAST_VISIT_DATE.format(form.schedule_form_id)
-            update_block.append(make_case_elem(last_visit_num))
 
-            self.add_setvalue(
-                ref='case/update/{}'.format(SCHEDULE_PHASE),
-                value=str(form.id + 1)
+            self.add_bind(
+                nodeset=u'/data/{}'.format(SCHEDULE_CURRENT_VISIT_NUMBER),
+                calculate=schedule_form_xpath.next_visit_due_num
             )
+            self.data_node.append(_make_elem(SCHEDULE_CURRENT_VISIT_NUMBER))
 
-            last_visit_prop_xpath = SESSION_CASE_ID.case().slash(last_visit_num)
-            self.add_setvalue(
-                ref='case/update/{}'.format(last_visit_num),
-                value="if({0} = '', 1, int({0}) + 1)".format(last_visit_prop_xpath)
+            self.add_bind(
+                nodeset=u'/data/{}'.format(SCHEDULE_UNSCHEDULED_VISIT),
+                calculate=schedule_form_xpath.is_unscheduled_visit,
+            )
+            self.data_node.append(_make_elem(SCHEDULE_UNSCHEDULED_VISIT))
```

Comment:
they can, but they have to type out the paths themselves (i.e. they don't show up as hidden variables)